networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -0,0 +1,716 @@
+=================================
+
+Implementation of approximate algorithms
+for solving and approximating the TSP problem.
+
+Categories of algorithms which are implemented:
+
+- Christofides (provides a 3/2-approximation of TSP)
+- Greedy
+- Simulated Annealing (SA)
+- Threshold Accepting (TA)
+
+The Travelling Salesman Problem tries to find, given the weight
+(distance) between all points where a salesman has to visit, the
+route so that:
+
+- The total distance (cost) which the salesman travels is minimized.
+- The salesman returns to the starting point.
+- Note that for a complete graph, the salesman visits each point once.
+
+The function `travelling_salesman_problem` allows for incomplete
+graphs by finding all-pairs shortest paths, effectively converting
+the problem to a complete graph problem. It calls one of the
+approximate methods on that problem and then converts the result
+back to the original graph using the previously found shortest paths.
+
+TSP is an NP-hard problem in combinatorial optimization,
+important in operations research and theoretical computer science.
+
+http://en.wikipedia.org/wiki/Travelling_salesman_problem
+"""
+import math
+import networkx as nx
+from networkx.utils import py_random_state, not_implemented_for
+
+__all__ = [
+    "traveling_salesman_problem",
+    "christofides",
+    "greedy_tsp",
+    "simulated_annealing_tsp",
+    "threshold_accepting_tsp",
+]
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+def christofides(G, weight="weight", tree=None):
+    """Approximate a solution of the traveling salesman problem
+
+    Compute a 3/2-approximation of the traveling salesman problem
+    in a complete undirected graph using Christofides [1]_ algorithm.
+
+    Parameters
+    ----------
+    G : Graph
+      Undirected complete graph
+
+    weight : string, optional (default="weight")
+      Edge data key corresponding to the edge weight.
+      If key not found, uses 1 as weight.
+
+    tree : NetworkX graph or None (default: computed)
+      A minimum spanning tree of G. Or, if None, the minimum spanning
+      tree is computed using :func:`networkx.minimum_spanning_tree`
+
+    Returns
+    -------
+    Generator
+        Generator of edges forming a 3/2-approximation of the minimal
+        Hamiltonian cycle.
+
+    References
+    ----------
+    .. [1] Christofides, Nicos. "Worst-case analysis of a new heuristic for
+       the travelling salesman problem." No. RR-388. Carnegie-Mellon Univ
+       Pittsburgh Pa Management Sciences Research Group, 1976.
+    """
+    # Check that G is a complete graph
+    for e in nx.selfloop_edges(G):
+        raise ValueError("Christofides algorithm does not allow selfloops")
+    n = len(G)
+    m = G.size()
+    if (n * (n - 1) // 2) != m:
+        raise ValueError("Christofides algorithm needs a complete graph")
+
+    if tree is None:
+        tree = nx.minimum_spanning_tree(G, weight=weight)
+    L = nx.Graph(G)
```

Comment:
The subtle difference between `nx.Graph(G)` and `G.copy()` is that if `G` is a multigraph, the first version converts it to a graph.  Here that isn't an issue because G is declared to be not directed and not multigraph.

So, I'll change it to G.copy() as you suggest. :}

Reply:
Got it, thanks. I wasn't sure what the differences were so thanks for the explanation :)